benchmark/coniql_performance_test.py

```diff
@@ -0,0 +1,341 @@
+import argparse
+import asyncio
+import datetime
+import json
+import sys
+import threading
+import time
+
+import psutil
+import websockets
+
+# Constants
+cpu_average = 0
+memory_use = 0
+PV_PREFIX = "TEST:REC"
+subscriptions_list = {}
+thread_list = []
+
+parser = argparse.ArgumentParser(description="Process inputs")
+parser.add_argument(
+    "-n", "--npvs", action="store", dest="n_pvs", default=1, help="Number of PVs"
+)
+parser.add_argument(
+    "-s",
+    "--nsamples",
+    action="store",
+    dest="n_samples",
+    default=10,
+    help="Number of samples to collect",
+)
+parser.add_argument(
+    "-p",
+    "--protocol",
+    action="store",
+    dest="ws_protocol",
+    choices=["1", "2"],
+    default=1,
+    help="websocket protocol: 1 = graphql-ws, 2 = graphql-transport-ws",
+)
+parser.add_argument(
+    "-f",
+    "--file",
+    action="store",
+    dest="output_file",
+    default="performance_test_results.txt",
+    help="File to output results to",
+)
+
+
+class GraphQLClient:
+    def __init__(self, endpoint, signal, ws_protocol):
+        self.endpoint = endpoint
+        self.signal = signal
+        self.ws_protocol = ws_protocol
+
+    async def subscribe(self, idid, query, handle, n_messages):
+        connection_init_message = json.dumps({"type": "connection_init", "payload": {}})
+
+        request_message_graphql_ws_protocol = json.dumps(
+            {"type": "start", "id": idid, "payload": {"query": query}}
+        )
+
+        request_message_graphql_transport_ws_protocol = json.dumps(
+            {"type": "subscribe", "id": idid, "payload": {"query": query}}
+        )
+
+        protocols = ["graphql-ws", "graphql-transport-ws"]
+
+        async with websockets.connect(
+            self.endpoint,
+            subprotocols=[protocols[self.ws_protocol - 1]],
+        ) as websocket:
+            await websocket.send(connection_init_message)
+            if self.ws_protocol == 2:
+                await websocket.send(request_message_graphql_transport_ws_protocol)
+            else:
+                await websocket.send(request_message_graphql_ws_protocol)
+
+            msg_count = 0
+            async for response in websocket:
```

Comment:
It'd be good to put a progress message somewhere in here - it's currently impossible to know how far through the test we are. Will need to do something funny with the output though, as this is in a thread the stdout will go elsewhere...